VMEncryption/main/EncryptionSettingsUtil.py

```diff
@@ -182,30 +184,49 @@ def _dict_to_name_value_array(self, values):
+                    else:
+                        kek_algorithm = CommonVariables.default_encryption_algorithm
+                        self.logger.log("No KEK algorithm specified, defaulting to {0}".format(kek_algorithm))
+
+        # ManagedHSM Checks
+        if key_store_type and key_store_type.lower() == CommonVariables.KeyStoreTypeManagedHSM.lower():
+            # validate mhsm parameters prior to creating the encryption settings object
+            if kv_url or kv_id:
+                raise Exception("KeyvaultUrl or KeyvaultresourceId are not empty, and 'KeyStoreType' parameter is set to ManagedHSM. Please remove KeyvaultUrl KeyvaultresourceId for ManagedHSM.")
+            self.logger.log("get_settings_data: KeyVault Url is empty, validating KeyEncryptionKeyKVURL and KeyEncryptionKeyKVId for ManagedHSM")
+            cutil.check_mhsm_url(kek_url, "A ManagedHSM URL is specified, but it is invalid for ManagedHSM.")
+            cutil.check_mhsm_id(kek_kv_id, "A ManagedHSM ID is required, but is missing or invalid.")
+            cutil.check_mhsm_name(kek_kv_id, kek_url, "A ManagedHSM ID and ManagedHSM URL were provided, but their ManagedHSM names did not match")
+            self._DISK_ENCRYPTION_KEY_SOURCE = CommonVariables.KeyStoreTypeManagedHSM
+        elif keystoretype_flag_exists:
+            raise Exception("The expected flag name and value to enable ManagedHSM is 'KeyStoreType':'{0}'. " +
+            "Please correct the flag name and value and retry enabling ManagedHSM, or remove the flag for KeyVault use.".format(CommonVariables.KeyStoreTypeManagedHSM))
+       # Keyvault Checks
         else:
-            if kek_kv_id:
-                raise Exception("The KEK KeyVault ID was specified but the KEK URL was missing")
+            # validate key vault parameters prior to creating the encryption settings object
+            cutil.check_kv_id(kv_id, "A KeyVault ID is required, but is missing or invalid")
+            cutil.check_kv_url(kv_url, "A KeyVault URL is required, but is missing or invalid")
+            cutil.check_kv_name(kv_id, kv_url, "A KeyVault ID and KeyVault URL were provided, but their key vault names did not match")
+            if kek_url:
+                cutil.check_kv_id(kek_kv_id, "A KEK URL was specified, but its KEK KeyVault ID was missing or invalid")
+                cutil.check_kek_url(kek_url, "A KEK URL was specified, but it was invalid")
+                cutil.check_kek_name(kek_kv_id, kek_url, "A KEK ID and KEK URL were provided, but their key vault names did not match")
+            else:
+                if kek_kv_id:
+                    raise Exception("The KEK KeyVault ID was specified but the KEK URL was missing")
 
         # create encryption settings object
         self.logger.log("Creating encryption settings object")
+        self.logger.log("The disk encryption key source used is: {0}".format(self._DISK_ENCRYPTION_KEY_SOURCE))
```

Comment:
This shouldn't matter but is there any reason there is no statement for keyvault source to initialize keyvaulttype?

`self._DISK_ENCRYPTION_KEY_SOURCE = CommonVariables.KeyStoreTypeKeyVault` 

Your logic would probably work because you are checking the value whether it is `CommonVariables.KeyStoreTypeManagedHSM` to decide the version.

But its just a comment for readability . I am okay without having this too

Reply:
I may have forgot to set that in the KV scenario, since we would remove it after preview. That should still be set though so that we have accurate data to analyze, I will ensure that KeyVault is set in the right case. Thanks